lib/std/array_list.zig

```diff
@@ -306,17 +307,10 @@ pub fn ArrayListAligned(comptime T: type, comptime alignment: ?u29) type {
         /// Invalidates pointers if additional memory is needed.
         pub fn ensureTotalCapacity(self: *Self, new_capacity: usize) Allocator.Error!void {
             if (@sizeOf(T) > 0) {
-                var better_capacity = self.capacity;
-                if (better_capacity >= new_capacity) return;
-
-                while (true) {
-                    better_capacity += better_capacity / 2 + 8;
-                    if (better_capacity >= new_capacity) break;
-                }
-
-                return self.ensureTotalCapacityPrecise(better_capacity);
+                if (self.capacity >= new_capacity) return;
+                return self.ensureTotalCapacityPrecise(math.max(8, new_capacity));
```

Comment:
Removing the `better_capacity` calculation here is not an acceptable change. The point of `better_capacity` is make the ArrayList grow such that the number of allocations is minimized (i.e. calling `ensureTotalCapacity` in a loop won't cause a reallocation on every call).

Here's how capacity grew before:

```
current capacity => resulting capacity
1...8 => 8
9...20 => 20
21...38 => 38
39...65 => 65
66...105 => 105
106...165 => 165
166...255 => 255
256...390 => 390
391...593 => 593
594...897 => 897
898...1353 => 1353
1354...2037 => 2037
2038...3063 => 4096
3064...6911 => 8192
6912...10374 => 12288
10375...15569 => 16384
15570...23361 => 24576
23362...35049 => 36864
35050...52581 => 53248
52582...78879 => 81920
78880...118326 => 118784
118327...177497 => 180224
177498...266253 => 270336
```

Any changes to this code should keep the same growth pattern or have benchmarks for why the new growth pattern is better in terms of real use cases.

Reply:
> The point of better_capacity is make the ArrayList grow such that the number of reallocations is minimized

i totally missed this but cannot understand why allocating the exact size (new way) is faster

Reply:
Ah, sorry, didn't notice that you edited the comment with the benchmarks. That is indeed interesting, I'd like to investigate that a bit more when I get a chance.

Reply:
forgot benchmark. i replaced `append` with `appendAssumeCapacity` and now, `new_way` looks super slow. i still wonder how it were faster, even with double calling it. at the end i think we'll go with **Saturating Arithmetic**

Reply:
an important thing. i made a mistake when comparing the binaries. and i thought this was even better so edited the real `new_way` and replaced with this (which is super slow):
```zig
return self.ensureTotalCapacityPrecise(std.math.max(8, new_capacity));
```
this is the real `new_way`:
```
return self.ensureTotalCapacityPrecise(std.math.max3(8, self.capacity * 2, new_capacity));
```

Reply:
I think you have to be careful how you construct the benchmarks for something like this. Your current benchmark code will always favor anything that grows the fastest (and your new_way grows faster than the current implementation). In fact, the `better_capacity` calculation that would perform the fastest in your benchmark would just be:

```zig
better_capacity = 100_000_000
```

since that would just do one allocation already at the final capacity requested in the benchmark, but it's obviously a very bad `better_capacity` calculation for the average case. In general, `better_capacity` is a tradeoff between:
- Number of reallocations needed
- Amount of memory allocated that is never actually used

I think the ideal is that `better_capacity` is tuned so that both are as low as possible in "real" usage.

So, I think some more sophisticated benchmarking would be necessary to evaluate different `better_capacity` algorithms (maybe something like [the self-hosted-parser benchmarks from gotta-go-fast](https://github.com/ziglang/gotta-go-fast/tree/master/benchmarks/self-hosted-parser) which makes heavy use of `ensureTotalCapacity`).

I think I'd again recommend restricting this PR to just solving the original integer overflow bug, and then creating a new issue/PR focused on the performance of the `better_capacity` algorithm.